<!doctype html>

<html lang="en" class="h-100">
  <head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
  <meta name="generator" content="Hugo 0.79.1" />
  <link rel="stylesheet" href="https://www.bustroker.com/css/bootstrap.min.css">
  
  
  <title>Handling parallel exceptions in dotnet | Coding notes</title>
  <style>
.container {
  max-width: 700px;
}
#nav a {
  font-weight: bold;
  color: inherit;
}
#nav a.nav-link-active {
  background-color: #212529;
  color: #fff;
}
#nav-border {
  border-bottom: 1px solid #212529;
}
#main {
  margin-top: 1em;
  margin-bottom: 4em;
}
#home-jumbotron {
  background-color: inherit;
}
#footer .container {
  padding: 1em 0;
}
#footer a {
  color: inherit;
  text-decoration: underline;
}
.font-125 {
  font-size: 125%;
}
.tag-btn {
  margin-bottom: 0.3em;
}
pre {
  background-color: #f5f5f5;
  border: 1px solid #ccc;
  border-radius: 4px;
  padding: 16px;
}
pre code {
  padding: 0;
  font-size: inherit;
  color: inherit; 
  background-color: transparent;
  border-radius: 0;
}
code {
  padding: 2px 4px;
  font-size: 90%;
  color: #c7254e;
  background-color: #f9f2f4;
  border-radius: 4px;
}
img,
iframe,
embed,
video,
audio {
  max-width: 100%;
}
.card-img,
.card-img-top,
.card-img-bottom {
  width: initial;
}
</style>
</head>
  <body class="d-flex flex-column h-100">
    <div id="nav-border" class="container">
  <nav id="nav" class="nav justify-content-center">
  
  
  
    
    
      
      
      
      
      
        
      
    
    
    <a class="nav-link " href="/"><i data-feather="home"></i> Home</a>
  
    
    
      
      
      
      
      
        
      
    
    
    <a class="nav-link " href="/notes/"><i data-feather="edit"></i> Notes</a>
  
    
    
      
      
      
      
      
        
      
    
    
    <a class="nav-link " href="/tags/"><i data-feather="tag"></i> Tags</a>
  
  </nav>
</div>

    <div class="container">
      <main id="main">
        

<h1>Handling parallel exceptions in dotnet</h1>
<p>I&rsquo;m executing several tasks in parallel, where more than one throw exceptions. I want to handle <strong>all the exceptions</strong>, which basically means catching them <strong>all</strong> and being able to, at least, rethrowing them with all the info they carry.</p>
<p>So, here it is how</p>
<div class="highlight"><pre style="color:#f8f8f2;background-color:#272822;-moz-tab-size:4;-o-tab-size:4;tab-size:4"><code class="language-c#" data-lang="c#"><span style="color:#66d9ef">var</span> runnersTasks = <span style="color:#66d9ef">new</span> List&lt;Task&gt;();<span style="color:#75715e">// list to keep the tasks created
</span><span style="color:#75715e"></span>Task compositeTask = <span style="color:#66d9ef">null</span>; <span style="color:#75715e">// task consisting in waiting for all the tasks to finish in parallel
</span><span style="color:#75715e"></span><span style="color:#66d9ef">try</span>
{
    <span style="color:#75715e">// run all the tasks and keep them in the list
</span><span style="color:#75715e"></span>    Runners.ForEach(r =&gt; runnersTasks.Add(r.RunAsync()));

    <span style="color:#75715e">// just wait for them all to finish, and keep the reference
</span><span style="color:#75715e"></span>    compositeTask = Task.WhenAll(runnersTasks);
    <span style="color:#66d9ef">await</span> compositeTask;
}
<span style="color:#66d9ef">catch</span> (Exception) <span style="color:#75715e">// here if you actually catch the exception, you&#39;ll find it&#39;s only the first one thrown
</span><span style="color:#75715e"></span>{
    <span style="color:#75715e">//usually just throw back the full AggregateException, rather than just the first exception catched here
</span><span style="color:#75715e"></span>    <span style="color:#66d9ef">throw</span> compositeTask.Exception;
}
</code></pre></div><h3 id="the-cathed-exception-is-only-the-first-one-thrown">The cathed exception is only the first one thrown.</h3>
<p>That&rsquo;s right. The list of all the exceptions is actually in the AggregateException kept in the property <em>Exception</em> of the <em>compositeTask</em>.</p>
<p>So, yes, the handling is that ugly.</p>
<h3 id="talk-is-cheap">Talk is cheap</h3>
<p><a href="https://github.com/bustroker/Bustroker.ParallelExceptions">Here&rsquo;s</a> the code.</p>



      </main>
    </div>
    
<footer id="footer" class="mt-auto text-center text-muted">
  <div class="container">
    Bustroker - Made with <a href="https://gohugo.io/">Hugo</a> &amp; <a href="https://github.com/zwbetz-gh/vanilla-bootstrap-hugo-theme">Vanilla</a>
  </div>
</footer>

    <script src="https://www.bustroker.com/js/feather.min.js"></script>
<script>
  feather.replace()
</script>


    


<link
  rel="stylesheet"
  href="https://cdn.jsdelivr.net/npm/katex@0.11.1/dist/katex.min.css"
  integrity="sha384-zB1R0rpPzHqg7Kpt0Aljp8JPLqbXI3bhnPWROx27a9N0Ll6ZP/+DiW/UqRcLbRjq"
  crossorigin="anonymous"
/>


<script
  defer
  src="https://cdn.jsdelivr.net/npm/katex@0.11.1/dist/katex.min.js"
  integrity="sha384-y23I5Q6l+B6vatafAwxRu/0oK/79VlbSz7Q9aiSZUvyWYIYsd+qj+o24G5ZU2zJz"
  crossorigin="anonymous"
></script>


<script
  defer
  src="https://cdn.jsdelivr.net/npm/katex@0.11.1/dist/contrib/auto-render.min.js"
  integrity="sha384-kWPLUVMOks5AQFrykwIup5lo0m3iMkkHrD0uJ4H5cjeGihAutqP0yW0J6dpFiVkI"
  crossorigin="anonymous"
  onload="renderMathInElement(document.body);"
></script>


    
  

  </body>
</html>